Introducing approximation into the MPS simulation method

merav-aharoni opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
We allow using approximation when using the MPS simulation method.


### Details and comments
The idea of approximation is to limit the number of Schmidt coefficients in the lambda matrices of the MPS. This can be done either by a size threshold or by the number of coefficients.
The approximation is controlled by two parameters at the beginning of svd.cpp:
`APPROX_THRES` - we discard values in lambda whose norm is smaller than 
APPROX_THRES * norm(S[0]), where S[0] is the largest Schmidt coefficient.
`APPROX_LIMIT` - we discard values only if the size of lambda is greater than this value
These parameters are currently hard-coded, and for now can be changed only in the code.
We need to add a user interface for these parameters, and to experiment with this option.

merav-aharoni commented:
I removed the WIP - this is ready for review

merav-aharoni commented:
@chriseclectic , @atilag , this is ready for review.

chriseclectic commented on src/simulators/qasm/qasm_controller.hpp:
```diff
@@ -82,6 +82,17 @@ namespace Simulator {
  *      use unallocated CPU cores up to max_parallel_threads. [Default: 100]
  *
+* From MatrixProductState::State class
+ * - "mps_approximation_threshold" (double): Set the cutoff value for the
```
I think this should be called `"matrix_product_state_bond_truncation_threshold"`, and should always be used (with a default value) regardless of whether the bond dimension is also specified. If you don't want any approximation this can be set to 0 by default.

chriseclectic commented on src/simulators/qasm/qasm_controller.hpp:
```diff
@@ -82,6 +82,17 @@ namespace Simulator {
+ *      Schmidt coefficients. Numbers smaller than the threshold will be discarded.
+ *      [Default: 1e-16]
+ *
+ * - "mps_approximation_num_sv" (uint): Set the cutoff value for the number
```
Can you change the name of this option to `"matrix_product_state_bond_dimension"`, since that is (I believe) the standard naming convention in tensor network community. 


```
*-  "matrix_product_state_bond_dimension" (uint): Set the maximum size of the matrix product state bond dimensions. This sets a cutoff value for the number of Schmidt coeffs...
```

chriseclectic commented on src/simulators/qasm/qasm_controller.hpp:
```diff
@@ -82,6 +82,17 @@ namespace Simulator {
+ *      Schmidt coefficients. If there are more than this number of coefficient, 
+ *      the smallest values will be discarded.
+ *      [Default: none]
+ * Note that both approximation parameters must be set for approximation to take place.
```
I would think approximation can happen for either option being used, or both.

* One option is to truncate almost-zero schmidt coefficients.
* The other is to limit the maximum number of non-zero schmidt coefficients

chriseclectic commented on qiskit/providers/aer/backends/qasm_simulator.py:
```diff
@@ -192,6 +192,31 @@ class QasmSimulator(AerBackend):
+      discarded (see approx_thresold).
+      (Default: NONE).
+
+    * ``"max_num_coefficients_for_approx"`` (int): Sets a limit on the number of Schmidt
```
Can you call this `"matrix_product_state_max_bond_dimension"` (see comment below in controller.hpp)

chriseclectic commented on qiskit/providers/aer/backends/qasm_simulator.py:
```diff
@@ -192,6 +192,31 @@ class QasmSimulator(AerBackend):
+      coefficients retained at the end of the svd algorithm.
+      If the number of coefficients is greater than max_num_coefficients_for_approx,
+      small values may be discarded, depending on approximation_type and on approx_threshold.
+      (Default: UINT_64MAX).
```
Isn't the default the full dimension of the n-qubit state space? Either way I think in documentation the default should be `Default: None`

chriseclectic commented on qiskit/providers/aer/backends/qasm_simulator.py:
```diff
@@ -192,6 +192,31 @@ class QasmSimulator(AerBackend):
+
+    * ``"approximation_type"`` (int): Can be one of the following options:
+      NO_APPROX - means no approximation.
+      ABSOLUTE_APPROX - only the largest max_num_coefficients_for_approx
```
I don't really understand what the different between these approximation types types are? It seems like we don't need this option, just two approximation options: a zero threshold, and a max bond dimension

chriseclectic commented on qiskit/providers/aer/backends/qasm_simulator.py:
```diff
@@ -192,6 +192,31 @@ class QasmSimulator(AerBackend):
+      small values may be discarded, depending on approximation_type and on approx_threshold.
+      (Default: UINT_64MAX).
+
+    * ``"approx_threshold"`` (double): Sets a lower limit on the value of the Schmidt
```
Same comment as for controller: I think this option should just be `"matrix_product_state_bond_truncation_threshold"` and should work independently of the max bond dimension truncation if set

merav-aharoni commented:
Hi @chriseclectic , thanks for the comments. I am fine with most of your comments (names of the parameters, updating the result metadata, and tests that were committed by mistake). 
Regarding the modes of approximation, I will try to explain my motivation. Often during the MPS algorithm, the arrays of Schmidt coefficients consists of identical entries. For example, we could get an array of length 64, where all entries are 0.0156. Using `truncation by max bond dimension` would remove an arbitrary number of entries, although these are no smaller than the others, so this might not be such a good approximation. Using a `truncation threshold of 0.01` would cause all entries to be removed, resulting in a crash. 
Therefore I introduced the third mode of approximation which is a more careful approximation: this is a `relative threshold` - i.e., percentage relative to the max entry in the array.
Let me know what you think, and if you still think I should remove the `relative threshold` or if I should modify it in some other way.

merav-aharoni commented:
@chriseclectic , I implemented the changes in MPS approximation as we discussed.
Regarding the `metadata`, I couldn't find any place where it is natural to add the specific metadata for a `State`. So I added a new method `State::add_metadata` in the base class. Every `State` can implement this method and add its own `metadata`.
This PR is ready for review.

chriseclectic commented on src/simulators/matrix_product_state/matrix_product_state.hpp:
```diff
@@ -375,6 +392,17 @@ void State::set_config(const json_t &config) {
+void State::add_metadata(ExperimentData &data, const json_t &config) const {
+  double threshold;
+  if (JSON::get_value(threshold, "matrix_product_state_truncation_threshold", config))
+    data.add_metadata("matrix_product_state_truncation_threshold", threshold);
```
These values could be stored as class variables, so you don't need to get them again from the json

atilag commented on src/simulators/matrix_product_state/svd.cpp:
```diff
@@ -95,11 +96,50 @@ uint_t num_of_SV(rvector_t S, double threshold)
+  // their squares is less than trunction_threshold
+  double sum_squares = 0;
+  for (int_t i=new_SV_num-1; i>0; i--) {
+    if (sum_squares + std::norm(S[i]) < truncation_threshold)
```
Can you put open and clone curly braces here?
```c++
{ 
}
```

atilag commented on src/simulators/matrix_product_state/svd.cpp:
```diff
@@ -95,11 +96,50 @@ uint_t num_of_SV(rvector_t S, double threshold)
+    }
+    if (1-sum > THRESHOLD) {
+      for (uint_t i=0; i<S.size(); i++) {
+	double square_i = std::norm(S[0])/sum;
+	S[i] = sqrt(square_i);
```
Can you indent this code?

atilag commented on src/simulators/matrix_product_state/svd.cpp:
```diff
@@ -523,7 +563,7 @@ status csvd(cmatrix_t &A, cmatrix_t &U,rvector_t &S,cmatrix_t &V)
 	  ss << "error: wrong SVD calc: A != USV*";
 	  throw std::runtime_error(ss.str());
-	}
+	  }
```
nit: remove the extra space.

atilag commented on src/simulators/state.hpp:
```diff
@@ -121,6 +121,14 @@ class State {
+  //-----------------------------------------------------------------------
+
+  // Every state can add information to the metadata structure
+  virtual void add_metadata(ExperimentData &data, const json_t &config) const {
```
This an empty method, we can just remove it, right?

atilag commented on src/simulators/matrix_product_state/matrix_product_state.hpp:
```diff
@@ -352,6 +360,15 @@ size_t State::required_memory_mb(uint_t num_qubits,
 void State::set_config(const json_t &config) {
+  // Set threshold for truncating Schmidt coefficients
+  double threshold;
+  if (JSON::get_value(threshold, "matrix_product_state_truncation_threshold", config))
```
nit: just for consistency I'd add curly braces to this conditional block

atilag commented on src/simulators/matrix_product_state/matrix_product_state.hpp:
```diff
@@ -375,6 +392,17 @@ void State::set_config(const json_t &config) {
 
+void State::add_metadata(ExperimentData &data, const json_t &config) const {
+  double threshold;
+  if (JSON::get_value(threshold, "matrix_product_state_truncation_threshold", config))
```
nit: Add curly braces here as well.

chriseclectic commented on src/simulators/matrix_product_state/svd.cpp:
```diff
@@ -28,9 +28,10 @@
 #define mul_factor 1e2
 #define tiny_factor 1e30
-#define THRESHOLD 1e-9
+#define THRESHOLD 1e-9 // threshold for re-normalization after approximation
```
Can these be made const vars or configurable vars rather than compile time defines?

chriseclectic commented:
@merav-aharoni Could you add a release note for this new feature?

merav-aharoni commented:
Added documentation for release notes.
